### PR TITLE
feat(transaction): Add Transaction Threshold model

### PR DESCRIPTION
**Pull request by shruthilayaj**
Introducing two models, ProjectTransactionThreshold that
will store custom Apdex/User Misery threshold on a project
level. ProjectTransactionThresholdOverride will allow user to
override the project level threshold with a transaction level
threshold.

### CODE REVIEW

File: src/sentry/models/transaction_threshold.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class TransactionThreshold(Model):
+    __core__ = False
+
+    # max_length here is based on the maximum for transactions in relay
+    transaction = models.CharField(max_length=200, null=True)
+    project = FlexibleForeignKey("sentry.Project", db_constraint=False)
+    organization = FlexibleForeignKey("sentry.Organization")
+    threshold = models.IntegerField()
+    metric = models.PositiveSmallIntegerField(default=TransactionMetric.DURATION.value)
```
Comment (Zylphrex) [inline src/sentry/models/transaction_threshold.py:22]: can we list all the allowed metrics as `choices=...` here?

Comment (shruthilayaj) [reply src/sentry/models/transaction_threshold.py:22]: I anticipate we'll be adding more metrics (for instance when we have a good mobile metric) and we'll need a migration every time we add new metrics if we added choices here. We can just handle the validation in the serializer I think and make sure to raise an error if it's not a valid choice.

File: src/sentry/models/transaction_threshold.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_transactionthreshold"
+        unique_together = (("project", "transaction"),)
+
+
+class ProjectTransactionThreshold(Model):
+    __core__ = False
+
+    project = FlexibleForeignKey("sentry.Project", unique=True)
```
Comment (Zylphrex) [inline src/sentry/models/transaction_threshold.py:33]: I believe we'll want to add `db_constraint=False` here as well

**Review by wedamija — COMMENTED** [review]
Just curious about how these will be used? The fields on `ProjectTransactionThreshold` look like a subset of the fields on `TransactionThreshold`

**Comment by shruthilayaj** [comment]
> Just curious about how these will be used? The fields on `ProjectTransactionThreshold` look like a subset of the fields on `TransactionThreshold`

We are going to allow users to set custom apdex/user misery thresholds on the project level (right now it's an org level setting) and for more fine grained control also let them set custom thresholds on the transaction level. Initially I tried to conflate the two into one model with the `transaction` field as a nullable column to set the project level threshold but this seemed slightly cleaner

**Comment by wedamija** [comment]
> > Just curious about how these will be used? The fields on `ProjectTransactionThreshold` look like a subset of the fields on `TransactionThreshold`
> 
> We are going to allow users to set custom apdex/user misery thresholds on the project level (right now it's an org level setting) and for more fine grained control also let them set custom thresholds on the transaction level. Initially I tried to conflate the two into one model with the `transaction` field as a nullable column to set the project level threshold but this seemed slightly cleaner

Ok, so `ProjectTransactionThreshold` is the primary threshold, and then `TransactionThreshold` is used to override?

Not totally sure if this is actually better, but just something to consider:

```
class ProjectTransactionThresholdOverride(Model):
    __core__ = False

    # max_length here is based on the maximum for transactions in relay
    project_transaction_threshold = FlexibleForeignKey("sentry.ProjectTransactionThreshold")
    transaction = models.CharField(max_length=200)
    threshold = models.IntegerField()
```

**Comment by shruthilayaj** [comment]
> Ok, so `ProjectTransactionThreshold` is the primary threshold, and then `TransactionThreshold` is used to override?
> 
> Not totally sure if this is actually better, but just something to consider:
> 
> ```
> class ProjectTransactionThresholdOverride(Model):
>     __core__ = False
> 
>     # max_length here is based on the maximum for transactions in relay
>     project_transaction_threshold = FlexibleForeignKey("sentry.ProjectTransactionThreshold")
>     transaction = models.CharField(max_length=200)
>     threshold = models.IntegerField()
> ```
> 
> Not sure on naming, but just this general concept.

Hmm I'm not sure what's better? Does the join provide a lot of value in this case? The main use case is we want to query the project defaults and the per-transaction overrides and build a snuba query based on those values. We might also want to look up the threshold for a single transaction which will mean always querying the Project level threshold table first and then checking if a transaction level override exists on the join. In the current models, we could just query the transaction threshold table first and exit early if an override exists

**Comment by wedamija** [comment]
> > Ok, so `ProjectTransactionThreshold` is the primary threshold, and then `TransactionThreshold` is used to override?
> > Not totally sure if this is actually better, but just something to consider:
> > ```
> > class ProjectTransactionThresholdOverride(Model):
> >     __core__ = False
> > 
> >     # max_length here is based on the maximum for transactions in relay
> >     project_transaction_threshold = FlexibleForeignKey("sentry.ProjectTransactionThreshold")
> >     transaction = models.CharField(max_length=200)
> >     threshold = models.IntegerField()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure on naming, but just this general concept.
> 
> Hmm I'm not sure what's better? Does the join provide a lot of value in this case? The main use case is we want to query the project defaults and the per-transaction overrides and build a snuba query based on those values. We might also want to look up the threshold for a single transaction which will mean always querying the Project level threshold table first and then checking if a transaction level override exists on the join. In the current models, we could just query the transaction threshold table first and exit early if an override exists

It reduces duplicated columns across the models, which is mostly helpful if you end up adding more columns to the model in the future. 

You should be able to do all the queries you want on this model - `ProjectTransactionThresholdOverride.objects.filter(project_transaction_threshold__project=<project>, project_transaction_threshold__transaction=<transaction>)` etc. The joins are cheap here since I assume this will be a pretty small table in general. One question though is whether you'd ever have a `TransactionThreshold` without a related `ProjectTransactionThreshold` row in your current design. If you do have that case, then my proposed design isn't that useful.

That said, this is just a suggestion and you know your use cases better than I do, so happy to go with either. If you go with the current design, I find the naming a little confusing as-is: It sounds like `ProjectTransactionThreshold` is more specific than `TransactionThreshold`, but it's the other way around, since `TransactionThreshold` is used as a more specific override for `ProjectTransactionThreshold`

**Comment by Zylphrex** [comment]
> It reduces duplicated columns across the models, which is mostly helpful if you end up adding more columns to the model in the future.
> 
> You should be able to do all the queries you want on this model - `ProjectTransactionThresholdOverride.objects.filter(project_transaction_threshold__project=<project>, project_transaction_threshold__transaction=<transaction>)` etc. The joins are cheap here since I assume this will be a pretty small table in general. One question though is whether you'd ever have a `TransactionThreshold` without a related `ProjectTransactionThreshold` row in your current design. If you do have that case, then my proposed design isn't that useful.
> 
> That said, this is just a suggestion and you know your use cases better than I do, so happy to go with either. If you go with the current design, I find the naming a little confusing as-is: It sounds like `ProjectTransactionThreshold` is more specific than `TransactionThreshold`, but it's the other way around, since `TransactionThreshold` is used as a more specific override for `ProjectTransactionThreshold`

One of the main reasons to go with the current design is exactly what you mentioned where it's possible to have a threshold set on the transaction but not on the project. So I think it makes sense to continue with this design.

I do agree that the naming is a little confusing so +1 to renaming them to something clearer.

**Comment by shruthilayaj** [comment]
> One of the main reasons to go with the current design is exactly what you mentioned where it's possible to have a threshold set on the transaction but not on the project. 

Yeah correct, we'd have to backfill all the project thresholds with the defaults if we wanted the join model to work.

Okay, I think it makes sense to stick with this too. I'll update the model name.

File: src/sentry/migrations/0196_add_project_transaction_threshold.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        migrations.CreateModel(
+            name="ProjectTransactionThreshold",
+            fields=[
+                (
+                    "id",
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                ("threshold", models.IntegerField()),
```
Comment (markstory) [inline src/sentry/migrations/0196_add_project_transaction_threshold.py:45]: Should this be a float?

File: src/sentry/migrations/0196_add_project_transaction_threshold.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                        on_delete=django.db.models.deletion.CASCADE, to="sentry.Organization"
+                    ),
+                ),
+                (
+                    "project",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        db_constraint=False,
+                        on_delete=django.db.models.deletion.CASCADE,
+                        to="sentry.Project",
+                        unique=True,
```
Comment (markstory) [inline src/sentry/migrations/0196_add_project_transaction_threshold.py:59]: What else is this unique with?

File: src/sentry/migrations/0196_add_project_transaction_threshold.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            fields=[
+                (
+                    "id",
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                ("transaction", models.CharField(max_length=200)),
+                ("threshold", models.IntegerField()),
+                ("metric", models.PositiveSmallIntegerField(default=1)),
```
Comment (markstory) [inline src/sentry/migrations/0196_add_project_transaction_threshold.py:78]: Should this be a choices based on the enum?

File: src/sentry/migrations/0196_add_project_transaction_threshold.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                    ),
+                ),
+                (
+                    "project",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        db_constraint=False,
+                        on_delete=django.db.models.deletion.CASCADE,
+                        to="sentry.Project",
+                    ),
+                ),
```
Comment (markstory) [inline src/sentry/migrations/0196_add_project_transaction_threshold.py:92]: Would it be useful to have `date_added` or an author on these models?

File: src/sentry/migrations/0196_add_project_transaction_threshold.py (hunk 3 above)
Comment (shruthilayaj) [reply src/sentry/migrations/0196_add_project_transaction_threshold.py:45]: We want to reduce the granularity of the threshold, only allow multiples of 100s, so we can just leave it as an integer field

File: src/sentry/migrations/0196_add_project_transaction_threshold.py (hunk 4 above)
Comment (shruthilayaj) [reply src/sentry/migrations/0196_add_project_transaction_threshold.py:59]: We just want it to be unique on the project- add/update one transaction threshold per project

Comment (markstory) [reply src/sentry/migrations/0196_add_project_transaction_threshold.py]: Ah ok. I didn't see the `unique_together` at the bottom 🤦

**Review by Zylphrex — APPROVED** [review]
LGTM

File: src/sentry/migrations/0197_add_project_transaction_threshold.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                ),
+                ("transaction", models.CharField(max_length=200)),
+                ("threshold", models.IntegerField()),
+                ("metric", models.PositiveSmallIntegerField(default=1)),
+                ("date_updated", models.DateTimeField(auto_now=True)),
+                (
+                    "edited_by",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True,
+                        on_delete=django.db.models.deletion.SET_NULL,
```
Comment (markstory) [inline src/sentry/migrations/0197_add_project_transaction_threshold.py:94]: Should this have `db_constraint=False` on it?

File: src/sentry/migrations/0197_add_project_transaction_threshold.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                ("threshold", models.IntegerField()),
+                ("metric", models.PositiveSmallIntegerField(default=1)),
+                ("date_updated", models.DateTimeField(auto_now=True)),
+                (
+                    "edited_by",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True,
```
Comment (markstory) [inline src/sentry/migrations/0197_add_project_transaction_threshold.py:52]: Should this have `db_constraint=False` on it?

Comment (shruthilayaj) [reply src/sentry/migrations/0197_add_project_transaction_threshold.py:52]: Oh yes, I suppose it should -- updating

Comment (shruthilayaj) [reply src/sentry/migrations/0197_add_project_transaction_threshold.py:53]: I forgot to push my changes 😑 , added in 6d26588bb1749bd6b2961f11228e56c7b384d265

File: src/sentry/models/transaction_threshold.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    __core__ = False
+
+    project = FlexibleForeignKey("sentry.Project", unique=True, db_constraint=False)
+    organization = FlexibleForeignKey("sentry.Organization")
+    threshold = models.IntegerField()
+    metric = models.PositiveSmallIntegerField(default=TransactionMetric.DURATION.value)
+    edited_by = FlexibleForeignKey(
+        "sentry.User", null=True, on_delete=models.SET_NULL, db_constraint=False
+    )
+    date_updated = models.DateTimeField(auto_now=True)
```
Comment (wedamija) [inline src/sentry/models/transaction_threshold.py:44]: Could actually use `DefaultFieldsModel` for both these models. It adds `date_updated` and `date_added` field for you.

Comment (shruthilayaj) [reply src/sentry/models/transaction_threshold.py:44]: Oh nice

**Review by markstory — COMMENTED** [review]
Looks good to me. Once the models are out, you should also think about how this model gets connected to deletions so that when projects or organizations are deleted all of this data is also removed.

**Review by wedamija — APPROVED** [review]
lgtm, could you update the pr description before merging?

**Pull request merged**